### DISCUSSION TITLE
How can i get the connect to Scrcpy video socket and control socket?

### DISCUSSION THREAD

**Discussion by Quynhquynh592** [yume-chan/ya-webadb#525] [General]
Hello my friend,
- When we connect to device, the server( android side) will return two socket, one for video stream and one for control device in one host "localhost:8000". ( i use adb forward to get this host)
Questions:
1. How can i connect exactly the socket?
Look like ex:
 let controller = socket.control; 
let video = socket.video;
2. If i get only the video stream, i can not control the device by mouse wheel, click or sth.. is it true?
3. I see in libraries has 
- adb-backend-direct-sockets
- adb-backend-webusb
- adb-backend-ws
- adb-credential-web
what is best way to connect device? Fast and smooth

If you have a code please provide it here, thank for your support.

**Comment by yume-chan** [comment]
> How can i connect exactly the socket?

How do you start the server?

> If i get only the video stream, i can not control the device by mouse wheel, click or sth.. is it true?

If you don't accept the control socket, I believe the server will stuck, timeout, throw an exception, and crash.

> what is best way to connect device? Fast and smooth

The first three are for connecting to devices using different transportation (TCP, WebSocket and USB), except the WebSocket one is unreliable (since WebSocket itself is unreliable), the other two are both fast and smooth. The credential pacakae is for managaing ADB keys.

**Comment by Quynhquynh592** [comment]
1. I push file scrcpy-server to device, and run "adb forward tcp:8000  localabstract:scrcpy" , " adb shell CLASSPATH...."
2. localhost:8000 return 2 socket, how can i know what is video socket or control socket?
I think it is look like object as {video: videoSocket; control: controlSocket}, but i dont know how to prove it. 
3. What is your way to split and parse data from video socket?
 Can you give me file link that you handle it before decode in your demo?
Thank bro...

**Reply by yume-chan** [reply]
When using forward connection, you need to connect to tcp://localhost:8000 twice to get two sockets. The first one is video socket and the second one is control (unless control was disabled).

As of Scrcpy 2.0 there is also an audio socket:

https://github.com/yume-chan/ya-webadb/blob/3a3f9ec4348b2aa5b4597d9c5becfaaec76c02ab/libraries/scrcpy/src/adb/connection.ts#L113-L125

> What is your way to split and parse data from video socket?

https://github.com/yume-chan/ya-webadb/blob/3a3f9ec4348b2aa5b4597d9c5becfaaec76c02ab/libraries/scrcpy/src/adb/client.ts#L351-L376

`parseVideoStreamMetadata` and `createMediaStreamTransformer` are different between versions. Check each versions' options class for detail.

---

If you want to know how Scrcpy works, I recommend you read its source code directly. Tango supports multiple versions of Scrcpy servers, so it's not the easiest way to learn.

**Comment by Quynhquynh592** [comment]
Thank bro for answer me.
